gcc/rust/backend/rust-compile-base.cc

```diff
@@ -108,7 +115,94 @@ HIRCompileBase::setup_fndecl (tree fndecl, bool is_main_entry_point,
 	{
 	  handle_no_mangle_attribute_on_fndecl (fndecl, attr);
 	}
+      else if (is_proc_macro)
+	{
+	  handle_bang_proc_macro_attribute_on_fndecl (fndecl, attr);
+	}
+      else if (is_proc_macro_attribute)
+	{
+	  handle_attribute_proc_macro_attribute_on_fndecl (fndecl, attr);
+	}
+      else if (is_proc_macro_derive)
+	{
+	  handle_derive_proc_macro_attribute_on_fndecl (fndecl, attr);
+	}
+    }
+}
+
+static void
+handle_proc_macro_common (tree fndecl, const AST::Attribute &attr)
+{
+  DECL_ATTRIBUTES (fndecl) = tree_cons (get_identifier ("gccrs_proc_macro"),
+					NULL, DECL_ATTRIBUTES (fndecl));
+}
+
+void
+HIRCompileBase::handle_bang_proc_macro_attribute_on_fndecl (
+  tree fndecl, const AST::Attribute &attr)
+{
+  handle_proc_macro_common (fndecl, attr);
+  ctx->collect_bang_proc_macro (fndecl);
+}
+
+void
+HIRCompileBase::handle_attribute_proc_macro_attribute_on_fndecl (
+  tree fndecl, const AST::Attribute &attr)
+{
+  handle_proc_macro_common (fndecl, attr);
+  ctx->collect_attribute_proc_macro (fndecl);
+}
+
+static std::vector<std::string>
+get_attributes (const AST::Attribute &attr)
+{
+  std::vector<std::string> result;
+
+  rust_assert (attr.get_attr_input ().get_attr_input_type ()
+	       == Rust::AST::AttrInput::TOKEN_TREE);
+  const auto &tt
+    = static_cast<const AST::DelimTokenTree &> (attr.get_attr_input ());
+
+  // TODO: Should we rely on fixed index ? Should we search for the
+  // attribute tokentree instead ?
+  if (tt.get_token_trees ().size () > 3)
+    {
+      rust_assert (tt.get_token_trees ()[3]->as_string () == "attributes");
+
+      auto attributes = static_cast<const AST::DelimTokenTree *> (
+	tt.get_token_trees ()[4].get ());
+      auto &token_trees = attributes->get_token_trees ();
```

Comment:
I think it'd be nice to have a little bit of documentation here explaining why we choose 3 and 4. something showing what the attribute looks like and what the token vector looks like for example

Reply:
Fixed in [b5ca1e4](https://github.com/Rust-GCC/gccrs/pull/2603/commits/b5ca1e4d9d85c991283fd1425af36fa628c1eb19).